clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/caching/KubernetesV2ProviderSynchronizable.java

```diff
@@ -0,0 +1,195 @@
+
+    // we only want to initialize caching agents for new or updated accounts
+    Set<KubernetesNamedAccountCredentials> allAccounts =
+        ProviderUtils.buildThreadSafeSetOfAccounts(
+                accountCredentialsRepository,
+                KubernetesNamedAccountCredentials.class,
+                ProviderVersion.v2)
+            .stream()
+            .filter(account -> newAndChangedAccounts.contains(account.getName()))
+            .collect(Collectors.toSet());
+
+    synchronizeKubernetesV2Provider(allAccounts);
+  }
+
+  private List<String> synchronizeAccountCredentials() {
+    List<String> deletedAccounts = getDeletedAccountNames();
+    List<String> changedAccounts = new ArrayList<>();
+    List<String> newAndChangedAccounts = new ArrayList<>();
+
+    deletedAccounts.forEach(accountCredentialsRepository::delete);
+
+    kubernetesConfigurationProperties.getAccounts().stream()
+        .filter(a -> ProviderVersion.v2.equals(a.getProviderVersion()))
+        .forEach(
+            managedAccount -> {
+              KubernetesNamedAccountCredentials credentials =
+                  new KubernetesNamedAccountCredentials(
+                      managedAccount, kubernetesSpinnakerKindMap, credentialFactory);
+
+              AccountCredentials existingCredentials =
+                  accountCredentialsRepository.getOne(managedAccount.getName());
+
+              if (existingCredentials == null) {
+                // account didn't previously exist
+                newAndChangedAccounts.add(managedAccount.getName());
+              } else if (existingCredentials != null && !existingCredentials.equals(credentials)) {
+                // account exists but has changed
+                changedAccounts.add(managedAccount.getName());
+                newAndChangedAccounts.add(managedAccount.getName());
+              }
+
+              accountCredentialsRepository.save(managedAccount.getName(), credentials);
+            });
+
+    ProviderUtils.unscheduleAndDeregisterAgents(deletedAccounts, catsModule);
+    ProviderUtils.unscheduleAndDeregisterAgents(changedAccounts, catsModule);
+
+    return newAndChangedAccounts;
+  }
+
+  private List<String> getDeletedAccountNames() {
+    List<String> existingNames =
+        accountCredentialsRepository.getAll().stream()
+            .filter(
+                (AccountCredentials c) ->
+                    KubernetesCloudProvider.getID().equals(c.getCloudProvider()))
+            .filter((AccountCredentials c) -> ProviderVersion.v2.equals(c.getProviderVersion()))
+            .map(AccountCredentials::getName)
+            .collect(Collectors.toList());
+
+    List<String> newNames =
+        kubernetesConfigurationProperties.getAccounts().stream()
+            .map(KubernetesConfigurationProperties.ManagedAccount::getName)
+            .collect(Collectors.toList());
+
+    return existingNames.stream()
+        .filter(name -> !newNames.contains(name))
```

Comment:
Similar to the comment for the V1 provider, this is `O(N^2)` in number of accounts.